Add request timings and count histograms to telemetry. Closes #3552

jberryman opened the pull request:
We upload a set of accumulating timers and counters to track service
time for different types of operations, across several dimensions (e.g.
did we hit the plan cache, was a remote involved, etc.)

<!-- Thank you for submitting this PR! :) -->
<!-- Provide a general summary of your changes in the Title above ^, end with (close #<issue-no>) or (fix #<issue-no>) -->

### Description
<!-- The title might not be enough to convey how this change affects the user. -->
<!-- Describe the changes from a user's perspective -->

### Affected components
<!-- Remove non-affected components from the list -->

- [x] Server
- [x] Tests

### Related Issues
#3552 although metrics should hopefully give us general visibility into performance

### Solution and Design
<!-- How is this issue solved/fixed? What is the design? -->
<!-- It's better if we elaborate -->

### Steps to test and verify
run tests

### Limitations, known bugs & workarounds
- marked an unsafe forked thread as `TODO immortal`, but we should fix all of these together (see #3768 )
- I didn't document changes in `telemetry.rst`; that doc is already stale. It would be nice if we could generate it from tests

### Server checklist
<!-- A checklist for server code -->

#### Catalog upgrade
<!-- Is hdb_catalog version bumped? -->
Does this PR change Hasura Catalog version?
- [x] No
- [ ] Yes
  - [ ] Updated docs with SQL for downgrading the catalog <!-- https://docs.hasura.io/1.0/graphql/manual/deployment/downgrading.html#downgrading-across-catalogue-versions -->

#### Metadata
<!-- Hasura metadata changes -->

Does this PR add a new Metadata feature?
- [x] No
- [ ] Yes
  - Does `run_sql` auto manages the new metadata through schema diffing?
    - [ ] Yes
    - [ ] Not required
  - Does `run_sql` auto manages the definitions of metadata on renaming?
    - [ ] Yes
    - [ ] Not required
  - Does `export_metadata`/`replace_metadata` supports the new metadata added?
    - [ ] Yes
    - [ ] Not required


#### GraphQL
- [x] No new GraphQL schema is generated
- [ ] New GraphQL schema is being generated:
   - [ ] New types and typenames are correlated
   <!-- No dangling types or typenames with missing types (a potential bug, introspection fails) -->
   <!-- If you have anything in your mind, which can be added here as a check list item, please submit a PR to update this template :) -->

#### Breaking changes

- [x] No Breaking changes
- [ ] There are breaking changes:

  1. Metadata API

     Existing `query` types:
     - [ ] Modify `args` payload which is not backward compatible
     - [ ] Behavioural change of the API
     - [ ] Change in response `JSON` schema
     - [ ] Change in error code
     <!-- Add if anything not listed above -->

  2. GraphQL API

     Schema Generation:
     <!-- Any changes in schema auto-generation logic -->
     <!-- All GraphQL schema names are case sensitive -->
     - [ ] Change in any `NamedType`
     - [ ] Change in table field names
     <!-- Add if anything not listed above -->

     Schema Resolve:-
     <!-- Any change in logic of resolving input request -->
     - [ ] Change in treatment of `null` value for any input fields <!-- Explain them below -->
     <!-- Add if anything not listed above -->

  3. Logging

     - [ ] Log `JSON` schema has changed
     - [ ] Log `type` names have changed
     <!-- Add if anything not listed above -->

<!-- Add any other breaking change not mentioned above -->

<!-- Explain briefly about your breaking changes below -->

jberryman commented:
@0x777 
I decided to base the resolution to #3547 on the work mentioned in the body of this PR to use some of the timing utilities. Can you please review 29b2963 ?

0x777 commented on server/src-lib/Hasura/Server/Logging.hs:
```diff
@@ -192,7 +191,10 @@ data OperationLog
   { olRequestId          :: !RequestId
   , olUserVars           :: !(Maybe UserVars)
   , olResponseSize       :: !(Maybe Int64)
+  , olQueryIoTime        :: !(Maybe Double)
```
What is `query_io_time`? From what I followed, is it the time taken for the request body to be read?

jberryman replied:
Yep, can the comment be improved?

I'm towards the end of a big refactor and rebase though, so some details might change. A question: do you think it's important to also include the IO wait time for the send of the response to the client? For whatever reason I didn't think to do this, but since there are often different bandwidth limits on uploads and downloads it might be a good idea. 

What do you think?

jberryman replied:
I couldn't see an obvious way to time sending the response to the client so haven't done that below. I can dig further into spock if that seems important.

jberryman commented:
@lexi-lambda @0x777 this is ready for review again

lexi-lambda commented on server/src-lib/Control/Concurrent/Extended.hs:
```diff
@@ -8,9 +10,15 @@ import           Prelude
+
+{-# DEPRECATED threadDelay "Please use `sleep` instead (and read the docs!)" #-}
+threadDelay :: Int -> IO ()
+threadDelay = Base.threadDelay
```
I understand the intent here (of guiding people who might have used `threadDelay` towards the new thing), but it seems a little curious to have a `DEPRECATED` binding in our own application code. Maybe it’s fine, though, I don’t have enormously strong feelings about it.

0x777 replied:
@jberryman Might as well get rid of `threadDelay` in the codebase?

jberryman replied:
In the last call I tried to outline the pattern. When (if) eventually this `Extended` module can shadow the standard module I think we should really strive to keep the changes and tweaks as obvious and discoverable as possible. I anticipate more `threadDelay` being written before the mixin shadowing thing happens, then all of these will become visible with a warning. I think there's no downside to keeping this in since we're fairly disciplined about keeping the haskell code warning free.

If I can do the mixin shadowing pattern successfully (now that we've moved to cabal-install) I'll write up a blog about the pattern and articulate thoughts a little more clearly and get feedback.

lexi-lambda commented on server/src-lib/Data/Time/Clock/Units.hs:
```diff
@@ -25,56 +25,72 @@ You can also go the other way using the constructors rather than the selectors:
   , Microseconds(..)
   , Nanoseconds(..)
+  -- * Reexports
+  -- | We use 'NominalDiffTime' as the standard type for duration in our code.
```
Could you add a comment about why we’ve chosen to use `NominalDiffTime` rather than `DiffTime` here, since the latter is technically more “correct”?

jberryman replied:
I'll add a comment, but I'm not totally sure what to say... it's not clear to me that one is more "correct" than the other as a measure of duration. I suppose `NominalDiffTime` indicates "may have been derived from sketchy math with `UTCTime`s" but that's not what we do

lexi-lambda replied:
`NominalDiffTime` is the correct type for *calendar* duration. It’s the appropriate type when you want to say “this event repeats every day,” since you don’t want it to drift earlier and earlier every time a leap second is introduced! It is not, however, an absolute measure of time.

`DiffTime`, on the other hand, *is* an absolute measure of time. This doesn’t make a significant difference if you’re measuring whole days, or even whole minutes, but if you’re measuring something that takes a second or two, the difference can be meaningful!

Of course, the datatypes don’t innately have any difference; they’re really just numbers. But they’re *supposed* to mean different things, which is why I think it’s surprising to use `NominalDiffTime` to measure wall clock time. I am not strongly opposed to using it this way, but I do think we should have a good justification for doing so.

Could you elaborate slightly on why you prefer using `NominalDiffTime`? It doesn’t seem like it would be enormously difficult to use `DiffTime`, but I assume you have a reason! I just want that reason to end up in the comment.

jberryman replied:
So the way you've described the distinction between `DiffTime` and `NominalDiffTime` is one way to squint at the situation but I don't think it's really reflected in the docs (they are at best ambiguous). In particular I don't think it's very meaningful to declare that `10 :: NominalDiffTime` is not an absolute measure of time. I think they are both _primarily_ durations, and the distinction between the two is only relevant as they relate to `UTCTime` (in the same way that newtype wrappers alter the semantics of a type wrt a particular operation), but again this depends on how you squint.

The comment I have is here https://github.com/hasura/graphql-engine/pull/3622/commits/179af277c5b011faf79cc626902c35db2d94deeb#diff-2ee44e54da5eae501804c04d09d0a270R48

I can't really think of something more interesting to say about this in a comment (but am open to suggestions!). Using `DiffTime` instead just means we have to convert from `DiffTime` to `NominalDiffTime` in various places where we use `addUTCTime` and `diffUTCTime` with no real benefit that I can see.

lexi-lambda replied:
What do you think is ambiguous about the documentation? It’s possible that there’s a reading of the documentation that I’m not seeing, but it seems fairly explicit:

> [`NominalDiffTime`] ignores leap-seconds, so it's not necessarily a fixed amount of clock time. For instance, 23:00 UTC + 2 hours of `NominalDiffTime` = 01:00 UTC (+ 1 day), regardless of whether a leap-second intervened.

1 hour in `NominalDiffTime` is one hour of calendar time, not 3,600 SI seconds (though admittedly the two are usually the same). For *measuring* elapsed time, that isn’t usually the type you want.

jberryman replied:
I think the docs are clumsily referring to the behavior of NominalDiffTime vis a vis UTCTime (which was the motivation for defining a new type as I understand it, it's a sin bin for sketchy time operations). 

In the context of an API in which the only way of constructing a `NominalDiffTime` was `diffUTCTime` (and no comparison operations were supported, and it couldn't be serialized, etc) I think you could say something abstract like "it's not necessarily a fixed amount of clock time" and really mean it.

lexi-lambda replied:
> I think the docs are clumsily referring to the behavior of NominalDiffTime vis a vis UTCTime (which was the motivation for defining a new type as I understand it, it's a sin bin for sketchy time operations).

I agree that this is technically true. There isn’t actually any representational difference between a `DiffTime` and a `NominalDiffTime`, so really the only difference is where the value comes from! It just so happens that you get a `NominalDiffTime` from `diffUTCTime`, which makes it calendar time. I’m in full agreement with you on all of that!

Still, I think the distinction is useful, even if it doesn’t necessarily add much in the way of actual type safety. Really, it’s much the same as writing something like `newtype UserName = UserName Text`. Sure, you *can* store arbitrary `Text` values in a `UserName` value, whether or not they’re actually usernames. But you shouldn’t!

So I think it does make sense to use `DiffTime` here. We can update `withElapsedTime` and `startTimer` to return `DiffTime` values without any additional difficulty, and we’ll get a small bit of type safety in return. If we ever find ourselves needing to deal with intervals of calendar time, we can revisit this—it wouldn’t be hard to generalize these types to be parameterized over the representation type so they work with both `DiffTime` and `NominalDiffTime`.

lexi-lambda commented on server/src-lib/Data/Time/Clock/Units.hs:
```diff
@@ -25,56 +25,72 @@ You can also go the other way using the constructors rather than the selectors:
+newtype Days = Days { days :: NominalDiffTime }
   deriving (Show, Eq, Ord)
-  deriving (Num, Fractional, Real, RealFrac) via (TimeUnit (SecondsP 86400))
+  deriving (Read, Num, Fractional, Real, RealFrac) via (TimeUnit (SecondsP 86400))
```
Do we really need the `Read` instances here? If so, why?

jberryman replied:
It was motivated from an application in this refactoring of a function like `readEnv :: Read a=> String -> IO a` (I can find the actually code if you need), but it seems generally useful for this kind of thing

lexi-lambda replied:
I don’t really care for `Read` as a typeclass, but I think this instance is especially sketchy, since it breaks the informal “law” that `Show` and `Read` instances are intended to roundtrip, if both are defined. This is not the case here, since we are deriving the stock `Show` instance but the underlying `Read` instance.

I’d prefer some slightly more explicit parsing logic. What about deriving [`ParseTime`](https://hackage.haskell.org/package/time-1.9.3/docs/Data-Time-Format.html#t:ParseTime), instead?

jberryman replied:
I'm fine removing the `Read` instance or implementing a `Show` that works in the same way

lexi-lambda replied:
I’d rather leave the `Read` instance off if there isn’t any code using it right now.

jberryman replied:
(as I said there is some code using it right now that motivated the instances, but that can be rewritten)

lexi-lambda replied:
Ah, I misunderstood if you were saying it was currently using it or if it was previously using it. If it’s currently using it, I won’t die on this hill; I’ll leave it up to you whether you want to change it or not.

The reason I don’t like `Read` is that I never have any idea what it does for any given type. Does it try to parse a Haskell expression that would produce the value? Does it use some custom serialization format? In the case of `readFloat` and other `Read` operations on numeric types, it goes and actually tries to parse a Haskell number literal, which I think is rarely what people intend!

Admittedly, in this particular instance, it probably isn’t so bad, so maybe it’s okay. I’m just always wary of relying on `Read`, since different instances work in wildly different ways.

lexi-lambda commented on server/src-lib/Hasura/GraphQL/Transport/HTTP.hs:
```diff
@@ -28,14 +31,23 @@ runGQ
-      E.execRemoteGQ reqId userInfo reqHdrs req rsi opDef
+  -- The response and misc telemetry data:
+  let telemTransport = Telem.HTTP
+  (telemTimeTot, (telemCacheHit, telemLocality, ((telemTimeIO, telemQueryType), !resp))) <- withElapsedTime $ do
```
Wouldn’t it be nice if Haskell had structurally typed records? 😭

jberryman replied:
yeah it wasn't very satisfying to write this stuff...

lexi-lambda commented on server/src-lib/Hasura/GraphQL/Transport/HTTP.hs:
```diff
@@ -72,10 +84,12 @@ runHasuraGQ
   -> UserInfo
   -> E.ExecOp
-  -> m EncJSON
+  -> m ((NominalDiffTime, Telem.QueryType), EncJSON)
```
It seems strange to me to use a nested pair here instead of a 3-tuple. Is there a particular reason to do it this way?

jberryman replied:
> Is there a particular reason to do it this way?

Laziness!

0x777 commented on server/src-lib/Hasura/Server/Logging.hs:
```diff
@@ -192,7 +191,10 @@ data OperationLog
   , olResponseSize       :: !(Maybe Int64)
-  , olQueryExecutionTime :: !(Maybe Double)
+  , olQueryIoTime        :: !(Maybe NominalDiffTime)
+  -- ^ Request IO wait time, i.e. time spent reading the full request from the socket.
```
Can the field name can be slightly more explicit (I feel that `query_io_time` suggests that this is the total time spent in io), maybe something like `request_read_time`?

0x777 replied:
How hard would it be implement the entire request time (https://github.com/hasura/graphql-engine/issues/3547)?

jberryman replied:
I agree that name is clearer

> How hard would it be implement the entire request time 

It seemed more useful to me to measure and report the network IO time directly, and you can add the two times together to get total (except for network IO time for sending the response) time. 

But if you are asking about how we can measure the network IO time for sending the response: this looked really difficult/invasive to do, since at the highest level we're passing spock continuations like `... -> Handler m (HttpResponse EncJSON)`. So it looked to me like we'd have to get into the internals of spock to do this, and it would be quite inconvenient. But I haven't looked into it deeply.

My thought was for the purposes of the underlying issue in #3547  (slow network issues being discoverable) measuring in one direction would be sufficient. Let me know what you think!

0x777 replied:
> But if you are asking about how we can measure the network IO time for sending the response: this looked really difficult/invasive to do, since at the highest level we're passing spock continuations like ... -> Handler m (HttpResponse EncJSON). So it looked to me like we'd have to get into the internals of spock to do this, and it would be quite inconvenient. But I haven't looked into it deeply

Fair enough, `request_read_time` should do for now.

jberryman commented:
@lexi-lambda I think this should be ready now.

lorenzo commented on server/src-lib/Hasura/Server/Telemetry/Counters.hs:
```diff
@@ -0,0 +1,183 @@
+
+-- | 'RequestTimings' along with the count
+data RequestTimingsCount  =
+  RequestTimingsCount {
```
The current model does not seem to enable the ability to calculate percentiles, which are extremely useful for performance tracking. Would you guys be open to explore an alternative design that would enable calculations beyond totals and averages?

jberryman replied:
For now for the purposes of telemetry we're just collecting counts and timings for a few hand chosen buckets, which should be fine for our purposes, rather than using hdrhistogram or something similar. But I agree with you that percentiles are useful!

jberryman commented:
@0x777 or @lexi-lambda 

This should address Alexis's changes (use DiffTime instead of NominalDiffTime). History got somewhat worse as I didn't want to spend the time to split up all the changes and squash them individually. I also added a few comments, added a `fromUnits` for conversions between units (including difftime and nominaldifftime), as I discovered `realToFrac` was a footgun.
